src/hist/plot.py

```diff
@@ -210,185 +220,386 @@ def plot2d_full(
+    hist_values = _hist.values()
+
+    # gaussian with reasonable initial guesses for parameters
+    constant = float(hist_values.max())
+    mean = (hist_values * x_values).sum() / hist_values.sum()
+    sigma = (hist_values * np.square(x_values - mean)).sum() / hist_values.sum()
+
+    def gauss(
+        x: np.ndarray,
+        constant: float = constant,
+        mean: float = mean,
+        sigma: float = sigma,
+    ) -> Any:
+        # Note: As return is a numpy ufuncs the type is "Any"
+        return constant * np.exp(-np.square(x - mean) / (2 * np.square(sigma)))
+
+    return gauss
+
+
+def _fit_callable_to_hist(
+    model: Callable[[np.ndarray], np.ndarray],
+    _hist: hist.BaseHist,
+    likelihood: bool = False,
+) -> "Tuple[np.ndarray, np.ndarray, np.ndarray, Tuple[Tuple[float, ...], np.ndarray]]":
+    """
+    Fit a model, a callable function, to the histogram values.
+    """
+    variances = _hist.variances()
+    if variances is None:
+        raise RuntimeError(
+            "Cannot compute from a variance-less histogram, try a Weight storage"
+        )
+    hist_uncert = np.sqrt(variances)
+
+    # Infer best fit model parameters and covariance matrix
+    xdata = _hist.axes[0].centers
+    popt, pcov = _curve_fit_wrapper(
+        model, xdata, _hist.values(), hist_uncert, likelihood=likelihood
+    )
+    model_values = model(xdata, *popt)
+
+    if np.isfinite(pcov).all():
+        n_samples = 100
+        vopts = np.random.multivariate_normal(popt, pcov, n_samples)
+        sampled_ydata = np.vstack([model(xdata, *vopt).T for vopt in vopts])
+        model_uncert = np.nanstd(sampled_ydata, axis=0)
+    else:
+        model_uncert = np.zeros_like(hist_uncert)
+
+    return model_values, model_uncert, hist_uncert, (popt, pcov)
+
+
+def _plot_fit_result(
+    _hist: hist.BaseHist,
+    model_values: np.ndarray,
+    model_uncert: np.ndarray,
+    ax: matplotlib.axes.Axes,
+    eb_kwargs: Dict[str, Any],
+    fp_kwargs: Dict[str, Any],
+    ub_kwargs: Dict[str, Any],
+) -> matplotlib.axes.Axes:
+    """
+    Plot fit of model to histogram data
+    """
+    x_values = _hist.axes[0].centers
+    variances = _hist.variances()
+    if variances is None:
+        raise RuntimeError(
+            "Cannot compute from a variance-less histogram, try a Weight storage"
+        )
+    hist_uncert = np.sqrt(variances)
+
+    _errorbars = ax.errorbar(x_values, _hist.values(), hist_uncert, **eb_kwargs)
+
+    # Ensure zorder draws data points above model
+    line_zorder = _errorbars[0].get_zorder() - 1
+    (line,) = ax.plot(x_values, model_values, **fp_kwargs, zorder=line_zorder)
+
+    # Uncertainty band for fitted function
+    # TODO: Probably set a better default color than the fit line color
+    ub_kwargs.setdefault("color", line.get_color())
+    ax.fill_between(
+        x_values,
+        model_values - model_uncert,
+        model_values + model_uncert,
+        **ub_kwargs,
+    )
+
+    return ax
```

Comment:
Generally, I'd return the Artists, not the Axes. The caller already has the axes, and you can get the axes from the Artists, too.

Reply:
This is probably not the right place, but I wonder if we should find a middle ground, maybe return the Axes in the named tuple as well. While it's easier to get axes from artists than the other way around, it makes for funny code and you want the axes more often than the artists.

Reply:
I think it's best to match matplotlib, and matplotlib always and only returns Artists. Most of the time, if you care about the axes, you already have them.

Reply:
In general, is there a more programmatic way to get the list (or I guess from below a NamedTuple) of Artists beyond manually bookkeeping? I assume that `ax.get_children()` is a bit too broad as I'm assuming you don't(?) want the `axis.XAxis` and `axis.YAxis` in there as well.

Reply:
oh. `plot2d_full` shows how: Just name the returns and of the plot calls and return them at the end. :+1:

Reply:
We are basically matching what matplotlib does. You can't get the exact Artists associated with a plotting function call except by the return. There could already be artists on the axis.

We are enhancing it slightly via named tuple, but same spirit.

Reply:
This was actually pretty educational in understanding the `matplotlib` API design better, or rather actually paying attenditon to what the API says it is doing. :grimacing: 

I still need to clean these up by using `namedtuple`, but this is hopefully slightly better shape now.